doc/man/rcat.1.txt
RCAT(1)
=======
:doctype: manpage

NAME
----
rcat - RDMA cat (netcat-alike using rdmacm/ibverb-compliant networking)

SYNOPSIS
--------
*rcat* {-s|-S addr|-c addr} [-p port] [-m] [-v] [-q] [-b blocksize] [-D statsprefix] [-d]

DESCRIPTION
-----------
rcat is a basic tool using RDMA send/recv with a very simple protocol. It probably cannot interface to another RDMA server/client directly, but it can at least be used to check that the connection establishes properly, check latency or naive bandwidth (see EXAMPLES below).

OPTIONS
-------
There is only one mandatory option, is rcat used as a client (and if so what's the server) or as a server? It can be one of the following:

*-c, --client* 'server'::
  We're a client, connect to 'server'. 'server' can be an IP address or hostname pointing to an IP.

*-s, --server*::
  We're a server, bind to all possible interfaces and IPs.

*-S 'IP'*::
  We're a server, but bind to a specific 'IP' (and only on the interface associated to it).

There are multiple optional options:

*-p, --port* 'port'::
  Port to use to bind or to connect to. 'port' can be either a numeric port or a service name, e.g. ``ssh'', resolvable by getaddrinfo. Defaults to 1235.

*-m, --multi*::
  Tell the server to accept more than one connection. A new thread will be spawned to handle each incoming connection.

*-v, --verbose*::
  Increase verbosity everytime it appears, e.g. -vvv will be more verbose than -v. Actually switches one more MSK_DEBUG_* flag on everytime.

*-q, --quiet*::
  Opposite operation.

*-b, --block-size* 'size'::
  Sets the block-size to 'size'. This is the maximum size allowed for a single RDMA send operation.

*-d*::
  Enables internal stats and prints them when the connection closes.

*-D, --stats* prefix::
  Enables stats and provide an unix socket interface to these, which can be used with `nc -u`. The socket name will be 'prefix'<threadID> (because a server can have multiple concurrent threads)

EXAMPLES
--------

.Bandwidth
Bandwidth can be checked with one or multiple dd and big chunks. rcat is _synchronous_, when a chunk is sent it will wait for the other side to receive the chunk, send an acknowledgement and will only sent the next chunk when it is received. This means that a single rcat client will never use the maximum bandwidth.

====
  server$ rcat -s -m > /dev/null
  client$ dd if=/dev/zero bs=1M count=4096 | rcat -c <dest>
4294967296 bytes (4.3 GB) copied, 3.25225 s, 1.3 GB/s
  client$ for i in `seq 1 4`; do dd if=/dev/zero bs=1M count=4096 | rcat -c <dest> & done; wait
4294967296 bytes (4.3 GB) copied, 5.17072 s, 831 MB/s
4294967296 bytes (4.3 GB) copied, 5.17226 s, 830 MB/s
4294967296 bytes (4.3 GB) copied, 5.17213 s, 830 MB/s
4294967296 bytes (4.3 GB) copied, 5.17315 s, 830 MB/s
====

.Latency
Lacency can be checked just the same with much smaller chunk. Keeping in mind that each chunk means a round-way trip, latency can be easily deducted as time/count.

----
  server$ rcat -s -m -b 10 > /dev/null
  client$ dd if=/dev/zero bs=10 count=100000 | rcat -c <dest> -b 10
1000000 bytes (1.0 MB) copied, 3.48947 s, 287 kB/s
-> 3.48947/100k = ~35us round trip
----

SEE ALSO
--------
mooshika(3), rmitm(1), rreplay(1)

AUTHORS
-------
mooshika and its tools have been written by Dominique Martinet <dominique.martinet@cea.fr>

RESOURCES
---------
https://github.com/martinetd/mooshika


src/rcat.c
/*
 *
 * Copyright CEA/DAM/DIF (2012)
 * contributor : Dominique Martinet  dominique.martinet@cea.fr
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * ---------------------------------------
 */

/**
 * \file   rcat.c
 * \brief  Example of usage/most basic test program for mooshika
 *
 * Example of usage/most basic test program for mooshika
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdio.h>	//printf
#include <stdlib.h>	//malloc
#include <string.h>	//memcpy
#include <unistd.h>	//read
#include <netdb.h>      //gethostbyname
#include <getopt.h>
#include <errno.h>
#include <poll.h>
#include <inttypes.h> // PRIu64


#include "utils.h"
#include "mooshika.h"

#define DEFAULT_BLOCK_SIZE 1024*1024
#define RECV_NUM 1

struct cb_arg {
	msk_data_t *ackdata;
	pthread_mutex_t *lock;
	pthread_cond_t *cond;
};

struct thread_arg {
	int mt_server;
	int stats;
	uint32_t block_size;
};

void callback_send(msk_trans_t *trans, msk_data_t *pdata, void *arg) {
}

void callback_disconnect(msk_trans_t *trans) {
	if (!trans->private_data)
		return;

	struct cb_arg *cb_arg = trans->private_data;
	pthread_mutex_lock(cb_arg->lock);
	pthread_cond_signal(cb_arg->cond);
	pthread_mutex_unlock(cb_arg->lock);
}


void callback_error(msk_trans_t *trans, msk_data_t *pdata, void *arg) {
	INFO_LOG(trans->state != MSK_CLOSING && trans->state != MSK_CLOSED
		&& trans->debug, "error callback on buffer %p", pdata);
}

void callback_recv(msk_trans_t *trans, msk_data_t *pdata, void *arg) {
	struct cb_arg *cb_arg = arg;
	int n;

	if (!cb_arg) {
		ERROR_LOG("no callback_arg?");
		return;
	}

	if (pdata->size != 1 || pdata->data[0] != '\0') {
	// either we get real data and write it to stdout/send ack
		n = write(1, (char *)pdata->data, pdata->size);
		fflush(stdout);

		if (n != pdata->size)
			ERROR_LOG("Wrote less than what was actually received");

		if (msk_post_recv(trans, pdata, callback_recv, callback_error, cb_arg))
			ERROR_LOG("post_recv failed");
	        if (msk_post_send(trans, cb_arg->ackdata, NULL, NULL, NULL))
			ERROR_LOG("post_send failed");
	} else {
	// or we get an ack and just send a signal to handle_thread thread
		if(msk_post_recv(trans, pdata, callback_recv, callback_error, cb_arg))
			ERROR_LOG("post_recv failed");

		pthread_mutex_lock(cb_arg->lock);
		pthread_cond_signal(cb_arg->cond);
		pthread_mutex_unlock(cb_arg->lock);
	}
}

void print_help(char **argv) {
	printf("Usage: %s {-s|-c addr} [-p port] [-m] [-v] [-b blocksize]\n", argv[0]);
	printf("Mandatory argument, either of:\n"
		"	-c, --client addr: client to connect to\n"
		"	-s, --server: server mode\n"
		"	-S addr: server mode, bind to address\n"
		"Optional arguments:\n"
		"	-p, --port port: port to use\n"
		"	-m, --multi: server only, multithread/accept multiple connections\n"
		"	-v, --verbose: enable verbose output (more v for more verbosity)\n"
		"	-q, --quiet: don't display connection messages\n"
		"	-D, --stats <prefix>: create a socket where to look stats up at given path\n"
		"	-d: display stats summary on close\n"
		"	-b, --block-size size: size of packets to send (default: %u)\n", DEFAULT_BLOCK_SIZE);
}

void* handle_trans(void *arg) {
	msk_trans_t *trans = arg;
	struct thread_arg *thread_arg = trans->private_data;
	uint8_t *rdmabuf;
	struct ibv_mr *mr;
	msk_data_t *wdata;
	msk_data_t *ackdata;

	pthread_mutex_t lock;
	pthread_cond_t cond;

	msk_data_t *rdata;
	struct cb_arg *cb_arg;
	int i;

	struct pollfd pollfd_stdin;


	// malloc memory zone that will contain all buffer data (for mr), and register it for our trans
#define RDMABUF_SIZE (RECV_NUM+2)*thread_arg->block_size
	TEST_NZ(rdmabuf = malloc(RDMABUF_SIZE));
	memset(rdmabuf, 0, RDMABUF_SIZE);
	TEST_NZ(mr = msk_reg_mr(trans, rdmabuf, RDMABUF_SIZE, IBV_ACCESS_LOCAL_WRITE));


	// malloc mooshika's data structs (i.e. max_size+size+pointer to actual data), for ack buffer
	TEST_NZ(ackdata = malloc(sizeof(msk_data_t)));
	ackdata->data = rdmabuf+(RECV_NUM+1)*thread_arg->block_size;
	ackdata->max_size = thread_arg->block_size;
	ackdata->size = 1;
	ackdata->data[0] = 0;
	ackdata->mr = mr;

	pthread_mutex_init(&lock, NULL);
	pthread_cond_init(&cond, NULL);

	// malloc receive structs as well as a custom callback argument, and post it for future receive
	TEST_NZ(rdata = malloc(RECV_NUM*sizeof(msk_data_t)));
	TEST_NZ(cb_arg = malloc(RECV_NUM*sizeof(struct cb_arg)));

	for (i=0; i < RECV_NUM; i++) {
		rdata[i].data=rdmabuf+i*thread_arg->block_size;
		rdata[i].max_size=thread_arg->block_size;
		rdata[i].mr = mr;
		cb_arg[i].ackdata = ackdata;
		cb_arg[i].lock = &lock;
		cb_arg[i].cond = &cond;
		TEST_Z(msk_post_recv(trans, &rdata[i], callback_recv, callback_error, &(cb_arg[i])));
	}

	trans->private_data = cb_arg;

	// receive buffers are posted, we can finalize the connection
	if (trans->server) {
		TEST_Z(msk_finalize_accept(trans));
	} else {
		TEST_Z(msk_finalize_connect(trans));
	}


	// malloc write (send) structs to post data read from stdin
	TEST_NZ(wdata = malloc(sizeof(msk_data_t)));
	wdata->data = rdmabuf+RECV_NUM*thread_arg->block_size;
	wdata->max_size = thread_arg->block_size;
	wdata->mr = mr;

	pollfd_stdin.fd = 0; // stdin
	pollfd_stdin.events = POLLIN | POLLPRI;
	pollfd_stdin.revents = 0;

	while (trans->state == MSK_CONNECTED) {

		i = poll(&pollfd_stdin, 1, 100);

		if (i == -1)
			break;

		if (i == 0)
			continue;

		wdata->size = read(0, (char*)wdata->data, wdata->max_size);
		if (wdata->size == 0)
			break;

		// post our data and wait for the other end's ack (sent in callback_recv)
		pthread_mutex_lock(&lock);
		TEST_Z(msk_post_send(trans, wdata, NULL, NULL, NULL));
		pthread_cond_wait(&cond, &lock);
		pthread_mutex_unlock(&lock);
	}	

	if (thread_arg->stats)
		printf("stats:\n"
			"	tx_bytes\ttx_pkt\n"
			"	%10"PRIu64"\t%"PRIu64"\n"
			"	rx_bytes\trx_pkt\n"
			"	%10"PRIu64"\t%"PRIu64"\n"
			"	error: %"PRIu64"\n"
			"	callback time:   %lu.%06lu s\n"
			"	completion time: %lu.%06lu s\n",
			trans->stats.tx_bytes, trans->stats.tx_pkt,
			trans->stats.rx_bytes, trans->stats.rx_pkt,
			trans->stats.err,
			trans->stats.nsec_callback / 1000000, trans->stats.nsec_callback % 1000000,
			trans->stats.nsec_compevent / 1000000, trans->stats.nsec_compevent % 1000000);


	TEST_Z(msk_dereg_mr(mr));

	msk_destroy_trans(&trans);

	// free stuff
	free(wdata);
	free(cb_arg);
	free(rdata);
	free(ackdata);
	free(rdmabuf);

	if (thread_arg->mt_server)
		pthread_exit(NULL);
	else
		return NULL;
}

int main(int argc, char **argv) {

	msk_trans_t *trans;
	msk_trans_t *child_trans;

	msk_trans_attr_t attr;

	struct thread_arg thread_arg;

	// argument handling
	static struct option long_options[] = {
		{ "client",	required_argument,	0,		'c' },
		{ "server",	no_argument,		0,		's' },
		{ "port",	required_argument,	0,		'p' },
		{ "help",	no_argument,		0,		'h' },
		{ "multi",	no_argument,		0,		'm' },
		{ "verbose",	no_argument,		0,		'v' },
		{ "quiet",	no_argument,		0,		'q' },
		{ "stats",	required_argument,	0,		'D' },
		{ "block-size",	required_argument,	0,		'b' },
		{ 0,		0,			0,		 0  }
	};

	int option_index = 0;
	int op;
	char *tmp_s;

	memset(&attr, 0, sizeof(msk_trans_attr_t));
	memset(&thread_arg, 0, sizeof(struct thread_arg));

	attr.server = -1; // put an incorrect value to check if we're either client or server
	// sane values for optional or non-configurable elements
	attr.debug = 1;
	attr.rq_depth = RECV_NUM+2;
	attr.disconnect_callback = callback_disconnect;
	attr.port = "1235"; /* default port */

	while ((op = getopt_long(argc, argv, "@hvqmsb:S:c:p:dD:", long_options, &option_index)) != -1) {
		switch(op) {
			case '@':
				printf("%s compiled on %s at %s\n", argv[0], __DATE__, __TIME__);
				printf("Release = %s\n", VERSION);
				printf("Release comment = %s\n", VERSION_COMMENT);
				printf("Git HEAD = %s\n", _GIT_HEAD_COMMIT ) ;
				printf("Git Describe = %s\n", _GIT_DESCRIBE ) ;
				exit(0);
			case 'b':
				thread_arg.block_size = strtoul(optarg, &tmp_s, 0);
				if (errno || thread_arg.block_size == 0) {
					thread_arg.block_size = 0;
					ERROR_LOG("Invalid block size, assuming default (%u)", DEFAULT_BLOCK_SIZE);
					break;
				}
				if (tmp_s[0] != 0) {
					set_size(&thread_arg.block_size, tmp_s);
				}
				INFO_LOG(attr.debug > 1, "block size: %u", thread_arg.block_size);
				break;
			case 'h':
				print_help(argv);
				exit(0);
			case 'v':
				attr.debug = attr.debug * 2 + 1;
				break;
			case 'd':
				thread_arg.stats = 1;
				break;
			case 'D':
				attr.stats_prefix = optarg;
				thread_arg.stats = 1;
				break;
			case 'q':
				attr.debug = 0;
				break;
			case 'c':
				attr.server = 0;
				attr.node = optarg;
				break;
			case 's':
				attr.server = 10;
				attr.node = "::";
				break;
			case 'S':
				attr.server = 10;
				attr.node = optarg;
				break;
			case 'p':
				attr.port = optarg;
				break;
			case 'm':
				thread_arg.mt_server = 1;
				break;
			default:
				ERROR_LOG("Failed to parse arguments");
				print_help(argv);
				exit(EINVAL);
		}
	}

	if (attr.server == -1) {
		ERROR_LOG("must be either a client or a server!");
		print_help(argv);
		exit(EINVAL);
	}

	if (thread_arg.block_size == 0)
		thread_arg.block_size = DEFAULT_BLOCK_SIZE;
	if (thread_arg.stats)
		attr.debug |= MSK_DEBUG_SPEED;

	// writing to stdout is the limiting factor anyway
	attr.worker_count = -1;

	TEST_Z(msk_init(&trans, &attr));

	if (!trans)
		exit(-1);


	trans->private_data = &thread_arg;

	if (trans->server) {
		pthread_t id;
		pthread_attr_t attr_thr;

		TEST_Z(msk_bind_server(trans));

		/* Init for thread parameter (mostly for scheduling) */
		if(pthread_attr_init(&attr_thr) != 0)
			ERROR_LOG("can't init pthread's attributes");

		if(pthread_attr_setscope(&attr_thr, PTHREAD_SCOPE_SYSTEM) != 0)
			ERROR_LOG("can't set pthread's scope");

		if(pthread_attr_setdetachstate(&attr_thr, PTHREAD_CREATE_DETACHED) != 0)
			ERROR_LOG("can't set pthread's join state");

		if (thread_arg.mt_server) {
			while (1) {
				child_trans = msk_accept_one(trans);
				if (!child_trans) {
					ERROR_LOG("accept_one failed!");
					break;
				}
				pthread_create(&id, &attr_thr, handle_trans, child_trans);
			}
		} else {
			child_trans = msk_accept_one(trans);
			handle_trans(child_trans);
		}
		msk_destroy_trans(&trans);
	} else { //client
		TEST_Z(msk_connect(trans));
		handle_trans(trans);
	       
	}

	return 0;
}
